nexus/src/sagas.rs

```diff
@@ -308,3 +335,332 @@ async fn sic_instance_ensure(
+    )
+    .await?;
+
+    Ok(region)
+}
+
+async fn sdc_regions_ensure(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let datasets_and_regions = sagactx
+        .lookup::<Vec<(db::model::Dataset, db::model::Region)>>(
+            "datasets_and_regions",
+        )?;
+    let request_count = datasets_and_regions.len();
+    futures::stream::iter(datasets_and_regions)
+        .map(|(dataset, region)| async move {
+            ensure_region_in_dataset(log, &dataset, &region).await
+        })
+        // Execute the allocation requests concurrently.
+        .buffer_unordered(request_count)
+        .collect::<Vec<Result<_, _>>>()
+        .await
+        .into_iter()
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(ActionError::action_failed)?;
+
+    // TODO: Region has a port value, we could store this in the DB?
+    Ok(())
+}
+
+async fn delete_regions(
+    datasets_and_regions: Vec<(db::model::Dataset, db::model::Region)>,
+) -> Result<(), Error> {
+    let request_count = datasets_and_regions.len();
+    futures::stream::iter(datasets_and_regions)
+        .map(|(dataset, region)| async move {
+            let url = format!("http://{}", dataset.address());
+            let client = CrucibleAgentClient::new(&url);
+            let id = RegionId(region.id().to_string());
+            client.region_delete(&id).await
+        })
+        // Execute the allocation requests concurrently.
+        .buffer_unordered(request_count)
+        .collect::<Vec<Result<_, _>>>()
+        .await
+        .into_iter()
+        .collect::<Result<Vec<_>, _>>()?;
+    Ok(())
+}
+
+async fn sdc_regions_ensure_undo(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<(), anyhow::Error> {
+    let datasets_and_regions = sagactx
+        .lookup::<Vec<(db::model::Dataset, db::model::Region)>>(
+            "datasets_and_regions",
+        )?;
+    delete_regions(datasets_and_regions).await?;
+    Ok(())
+}
+
+async fn sdc_finalize_disk_record(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let _params = sagactx.saga_params();
+
+    let disk_id = sagactx.lookup::<Uuid>("disk_id")?;
+    let disk_created = sagactx.lookup::<db::model::Disk>("created_disk")?;
+    osagactx
+        .datastore()
+        .disk_update_runtime(&disk_id, &disk_created.runtime().detach())
+        .await
+        .map_err(ActionError::action_failed)?;
+    Ok(())
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ParamsDiskDelete {
+    pub disk_id: Uuid,
+}
+
+#[derive(Debug)]
+pub struct SagaDiskDelete;
+impl SagaType for SagaDiskDelete {
+    type SagaParamsType = Arc<ParamsDiskDelete>;
+    type ExecContextType = Arc<SagaContext>;
+}
+
+fn saga_disk_delete() -> SagaTemplate<SagaDiskDelete> {
+    let mut template_builder = SagaTemplateBuilder::new();
+
+    template_builder.append(
```

Comment:
I thought it would be critical to modify the disk record before anything else - if we don't update the disk record first, couldn't other concurrent operations poke and prod at the disk while we're tearing down the backing storage?

In the "disk creation" saga, we include a final step to "finalize" the disk, which basically exposes it for access. I figured the most important invariant was "if a disk has been created, it should be backed by functioning regions at all time" - so taking it out of the rotation felt like the most important step, before taking out regions.

Reply:
I see -- yeah, and I think on the creation side, we use an intermediate state ("creating") to enforce both of these invariants.  I guess we could do that here too.  I'm not sure if it matters or not.  It still strikes me as a little strange to have undeleted resources associated with a deleted one.  But maybe it's not that useful to distinguish between `destroying` and `destroyed`.